apps/openmw/mwphysics/physicssystem.cpp

```diff
@@ -674,6 +688,876 @@ namespace MWPhysics
+            for (auto v : mObjects)
+                if (const auto concaveShape = dynamic_cast<btConcaveShape*>(v->getShapeInstance()->mCollisionShape))
+                    mMeshBuilder.addShape(*concaveShape, v->getCollisionObject()->getWorldTransform());
+        }
+    };
+
+// Use to dump scene to load from recastnavigation demo tool
+#ifdef OPENMW_WRITE_OBJ
+    void writeObj(const std::vector<float>& vertices, const std::vector<int>& indices)
+    {
+        const auto path = std::string("scene.") + std::to_string(std::time(nullptr)) + ".obj";
+        std::ofstream file(path);
+        if (!file.is_open())
+            throw NavigatorException("Open file failed: " + path);
+        file.exceptions(std::ios::failbit | std::ios::badbit);
+        file.precision(std::numeric_limits<float>::max_digits10);
+        std::size_t count = 0;
+        for (auto v : vertices)
+        {
+            if (count % 3 == 0)
+            {
+                if (count != 0)
+                    file << '\n';
+                file << 'v';
+            }
+            file << ' ' << v;
+            ++count;
+        }
+        file << '\n';
+        count = 0;
+        for (auto v : indices)
+        {
+            if (count % 3 == 0)
+            {
+                if (count != 0)
+                    file << '\n';
+                file << 'f';
+            }
+            file << ' ' << (v + 1);
+            ++count;
+        }
+        file << '\n';
+    }
+#endif
+
+    class RecastMeshManagerCache
+    {
+    public:
+        template <class T>
+        bool addObject(T* object)
+        {
+            if (!mImpl.addObject(object))
+                return false;
+            mCached.reset();
+            return true;
+        }
+
+        template <class T>
+        bool removeObject(T* object)
+        {
+            if (!mImpl.removeObject(object))
+                return false;
+            mCached.reset();
+            return true;
+        }
+
+        const RecastMesh& getMesh()
+        {
+            if (!mCached.is_initialized())
+                mCached = mImpl.getMesh();
+#ifdef OPENMW_WRITE_OBJ
+            writeObj(mCached->getVertices(), mCached->getIndices());
+#endif
+            return *mCached;
+        }
+
+    private:
+        RecastMeshManager mImpl;
+        boost::optional<RecastMesh> mCached;
+    };
+
+    struct AgentParams
+    {
+        float mHeight;
+        float mMaxClimb;
+        float mRadius;
+    };
+
+    std::tuple<float, float, float> makeTuple(const AgentParams& value)
+    {
+        return std::make_tuple(value.mHeight, value.mMaxClimb, value.mRadius);
+    }
+
+    struct LessAgentParams
+    {
+        bool operator ()(const AgentParams& lhs, const AgentParams& rhs) const
+        {
+            return makeTuple(lhs) < makeTuple(rhs);
+        }
+    };
+
+    using NavMeshPtr = std::unique_ptr<dtNavMesh, decltype(&dtFreeNavMesh)>;
```

Comment:
This causes the `unique_ptr` to have to hold two pointers. You can probably instead do:
```
/* Define unique_ptrs to auto-cleanup associated objects. */
struct NavMeshDeleter {
    void operator()(dtNavMesh *ptr) { dtFreeNavMesh(ptr); }
};
using NavMeshPtr = std::unique_ptr<dtNavMesh,NavMeshDeleter>;
```
which makes it only need to store the object pointer (the deleter type is known at compile time and it can call the `operator()` method, so it doesn't need any run-time storage to make a call to `dtFreeNavMesh`).